pkg/bgpv1/README.md

```diff
@@ -0,0 +1,306 @@
+# Cilium BGP Control Plane
+
+## Usage
+
+Currently a single flag in the `Cilium Agent` exists to turn on the `BGP Control Plane` feature set.
+
+```
+--enable-bgp-control-plane=true
+```
+
+When set to `true` the `BGP Control Plane` `Controllers` will be instantiated and will begin listening for `CiliumBGPPeeringPolicy` events.
+
+### CiliumBGPPeeringPolicy CRD
+
+All `BGP` peering topology information is carried in a `CiliumBGPPeeringPolicy` CRD. 
+
+`CiliumBGPPeeringPolicy` can be applied to one or more nodes based on its `nodeSelector` fields. 
+
+A Cilium node may only have a single `CiliumBGPPeeringPolicy` apply to it and if more then one does, it will apply no policy at all.
+
+Each `CiliumBGPPeeringPolicy` defines one or more `CiliumBGPVirtualRouter` configurations.
+
+When these CRDs are written or read from the cluster the `Controllers` will take notice and perform the necessary actions to drive the `BGP Control Plane` to the desired state described by the policy.
+
+The policy in `yaml` form is defined below:
+
+```yaml
+apiVersion: "cilium.io/v2alpha1"
+kind: CiliumBGPPeeringPolicy
+metadata:
+ name: 01-bgp-peering-policy
+spec: # CiliumBGPPeeringPolicySpec
+ nodeSelector:
+   matchLabels:
+     bgp-policy: a
+ virtualRouters: # []CiliumBGPVirtualRouter
+ - localASN: 64512
+   exportPodCIDR: true
+   neighbors: # []CiliumBGPNeighbor
+	 - peerAddress: 'fc00:f853:ccd:e793::50/128'
+	   peerASN: 64512
+```
+
+#### Fields
+``` 
+nodeSelector: Nodes which are selected by this label selector will apply the given policy
+ 
+ virtualRouters: One or more peering configurations outlined below. Each peering configuration can be thought of as a BGP router instance.
+	
+	virtualRouters[*].localASN: The local ASN for this peering configuration
+	
+	virtualRouters[*].exportPodCIDR: Whether to export the private pod CIDR block to the listed neighbors
+	
+	virtualRouters[*].neighbors: A list of neighbors to peer with
+		neighbors[*].peerAddress: The address of the peer neighbor
+		neighbors[*].peerASN: The ASN of the peer	
+```
+
+*note*: *setting unique configuration details of a particular instantiated virtual router on a particular Cilium node is explained in [Virtual Router Attributes](#Virtual Router Attributes)*
+
+## Creating  a BGP Topology
+
+### Rules
+
+The following rules must be adhered to for a `CiliumBGPPeeringPolicy` to correctly apply to a node.
+
+- Only a single `CiliumBGPPeeringPolicy` can apply to a `Cilium` node. 
+	* If the `BGP Control Plane` on a node iterates through the `CiliumBGPPeeringPolicy` CRs currently written to the cluster and discovers (n > 1) policies match its labels, it will return an error and remove any existing BGP sessions. Only (n == 1) policies **must** match a node's label sets.
```

Comment:
I have a concern for this behavior. If I read this sentence (and the code), when the user mistakenly applies the wrong configuration over the existing configuration, that immediately withdraws the route from the peer. So, they will experience immediate connectivity disasters.

How about leaving the existing BGP session as is when the user applies the new policy over the existing policy and reporting the error with log (or k8s resource event) to the users? In my opinion, that should be more comfortable for users.

Reply:
This was disccused with @joestringer during CFP process.

The idea was to fail quick in the user's staging environments so that the issue is known ahead of time and can be fixed. 

Down to revisit this behavior tho.

Reply:
Yeah my goal here is to say "if there is conflicting configuration, make it clear to the user as quickly as possible". One of the other background problems that I see is that if we attempt to allow one of the configurations to stay, and then we restart the Cilium agent, then we cannot deterministically figure out which configuration should apply to continue the connectivity that was present before the restart. So, instead of causing the user to understand the problem when they introduce the bad configuration, they may only find the problem upon Cilium restart or upgrade, at which point it is less clear why the BGP sessions suddenly failed. This would be more difficult to trace back to the origin of the problem (conflicting user configuration).

In an ideal world, maybe we could support multiple configurations at once. But also, I don't think we have specific use cases in mind for this just yet, and it is more complicated to implement such support. So I'm quite OK with this behaviour to withdraw the connections as soon as there is conflicting configuration from the user.